### PR TITLE
[draft] feat(portforward): Add SSHD module + Ability to port forward

### PR DESCRIPTION
**Pull request by JulienBreux**
## What does this PR do?

Rel: https://github.com/testcontainers/testcontainers-go/issues/2212

## Why is it important?

To call the host from a container inside a testcontainers container.
Related to Microcks stuff.
(because a Microcks async minion had to call a host development webserver for example)

## Related issues

- Relates #2212

## How to test this PR

Use `.WithAccessToHost(true)`

### CODE REVIEW

File: modules/sshd/options.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+// WithRootPassword sets the root password
+func WithRootPassword(password string) Option {
+	return func(o *options) {
+		o.RootPassword = password
+	}
+}
+
+// WithRootPassword sets the options
+func WithOptions(opts []string) Option {
```
Comment (JulienBreux) [inline modules/sshd/options.go:41]:
```suggestion
func WithCommandOptions(commandOption []string) Option {
```

File: modules/sshd/options.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	return func(o *options) {
+		o.Options = opts
```
Comment (JulienBreux) [inline modules/sshd/options.go:43]:
```suggestion
		o.CommandOptions = commandOptions
```

File: modules/sshd/sshd.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+type SSHDContainer struct {
+	testcontainers.Container
+}
+
+// RunContainer creates an instance of the SSHD container type
+func RunContainer(ctx context.Context, opts ...testcontainers.ContainerCustomizer) (*SSHDContainer, error) {
+	req := testcontainers.ContainerRequest{
+		Image:        defaultImage,
+		ExposedPorts: []string{defaultPort},
+		// WaitingFor:   wait.ForLog("* Ready to accept connections"),
```
Comment (JulienBreux) [inline modules/sshd/sshd.go:32]: ⚠️

File: modules/sshd/sshd.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	settings := defaultOptions()
+	for _, opt := range opts {
+		if apply, ok := opt.(Option); ok {
+			apply(&settings)
+		}
+		opt.Customize(&genericContainerReq)
+	}
+
+	applyRootPassword(settings.RootPassword)(&genericContainerReq)
+	applyCommandOptions(settings.Options)(&genericContainerReq)
```
Comment (JulienBreux) [inline modules/sshd/sshd.go:51]:
```suggestion
	applyCommandOptions(settings.CommandOptions)(&genericContainerReq)
```

**Comment by JulienBreux** [comment]
In progress log:
The first idea is to have a SSHD module in my design.
After a full compatible SSHD module, the second idea is to have a `PortForwarding` option into the main Container perspective to run this SSHD port forwarding system.

**Comment by mdelapenya** [comment]
Hi @JulienBreux our idea is to have the sshd container as part of the internals of the library, so the users just set, e.g., a field in the request struct with the list of ports needed by the container, and the library automatically spins up the sshd container under the hood, forwarding those ports to the container through the sshd server container.

```go
req := testcontainers.GenericContainerRequest{
		ContainerRequest: testcontainers.ContainerRequest{
			Image:           "alpine:3.17",
			HostAccessPorts: []int{3000, 8080},
		}
}
```

I'm currently working on that approach.

**Comment by JulienBreux** [comment]
> I'm currently working on that approach.

Hi, don't hesitate, if I can help.
My first idea was to create a module SSHD and then using it in a PortForwarding brick.
I'm waiting for you, no pb :)

**Comment by JulienBreux** [comment]
Closed by https://github.com/testcontainers/testcontainers-go/pull/2471
Removing contribution ✅

**Pull request closed without merging**